File: python/slhdsa/xmss.py
"""XMSS functions for SLH-DSA per FIPS 205, Section 6."""

from slhdsa.params import SLHDSAParams
from slhdsa.address import ADRS, TREE, WOTS_HASH
from slhdsa.hash_funcs import get_suite
from slhdsa.wots import wots_pkgen, wots_sign, wots_pk_from_sig


def xmss_node(params: SLHDSAParams, sk_seed: bytes, i: int, z: int,
              pk_seed: bytes, adrs: ADRS) -> bytes:
    """Compute internal XMSS node (Algorithm 8).

    Args:
        i: Node index at height z.
        z: Target height in the XMSS tree.
    """
    suite = get_suite(params.hash_type)

    if z > params.hp or i >= (1 << (params.hp - z)):
        return None

    if z == 0:
        # Leaf node: WOTS+ public key
        adrs.set_type(WOTS_HASH)
        adrs.set_key_pair_address(i)
        return wots_pkgen(params, sk_seed, pk_seed, adrs)
    else:
        # Internal node
        lnode = xmss_node(params, sk_seed, 2 * i, z - 1, pk_seed, adrs)
        rnode = xmss_node(params, sk_seed, 2 * i + 1, z - 1, pk_seed, adrs)

        adrs.set_type(TREE)
        adrs.set_tree_height(z)
        adrs.set_tree_index(i)
        return suite.H(pk_seed, adrs, lnode + rnode, params.n)


def xmss_sign(params: SLHDSAParams, m: bytes, sk_seed: bytes, idx: int,
              pk_seed: bytes, adrs: ADRS) -> bytes:
    """Generate XMSS signature (Algorithm 9).

    Args:
        m: n-byte message.
        idx: Leaf index to sign with.

    Returns:
        XMSS signature: WOTS+ sig || auth path.
    """
    # Build authentication path
    auth = b""
    for j in range(params.hp):
        # Sibling index at height j: flip bit j of idx, then shift right by j
        s = (idx >> j) ^ 1
        auth += xmss_node(params, sk_seed, s, j, pk_seed, adrs)

    # Generate WOTS+ signature on m
    adrs.set_type(WOTS_HASH)
    adrs.set_key_pair_address(idx)
    sig = wots_sign(params, m, sk_seed, pk_seed, adrs)

    return sig + auth


def xmss_pk_from_sig(params: SLHDSAParams, idx: int, sig_xmss: bytes,
                     m: bytes, pk_seed: bytes, adrs: ADRS) -> bytes:
    """Compute XMSS root from signature (Algorithm 10).

    Args:
        idx: Leaf index.
        sig_xmss: XMSS signature (WOTS sig || auth path).
        m: n-byte message that was signed.

    Returns:
        Computed XMSS root node.
    """
    suite = get_suite(params.hash_type)
    n = params.n

    # Split signature
    wots_sig = sig_xmss[:params.wots_len * n]
    auth = sig_xmss[params.wots_len * n:]

    # Compute WOTS+ public key from signature
    adrs.set_type(WOTS_HASH)
    adrs.set_key_pair_address(idx)
    node = wots_pk_from_sig(params, wots_sig, m, pk_seed, adrs)

    # Climb auth path to compute root
    adrs.set_type(TREE)
    k = idx  # current node index tracking
    for j in range(params.hp):
        adrs.set_tree_height(j + 1)
        auth_j = auth[j * n:(j + 1) * n]
        parent = k >> 1
        adrs.set_tree_index(parent)
        if k % 2 == 0:
            node = suite.H(pk_seed, adrs, node + auth_j, n)
        else:
            node = suite.H(pk_seed, adrs, auth_j + node, n)
        k = parent

    return node

File: js/src/slhdsa/xmss.js
/**
 * XMSS (eXtended Merkle Signature Scheme) for SLH-DSA (FIPS 205)
 * Algorithms 9-11
 */

import { ADRS, ADDR_TYPE } from './address.js';
import { wotsPkgen, wotsSign, wotsPkFromSig } from './wots.js';

/**
 * xmssNode(skSeed, pkSeed, i, z, adrs, params, hashSuite) - Algorithm 9
 * Compute the root of a subtree of height z at index i
 */
function xmssNode(skSeed, pkSeed, i, z, adrs, params, hashSuite) {
  const { hp, n } = params;

  if (z === 0) {
    // Leaf node: WOTS+ public key
    adrs.setType(ADDR_TYPE.WOTS_HASH);
    adrs.setKeyPairAddress(i);
    return wotsPkgen(skSeed, pkSeed, adrs, params, hashSuite);
  }

  const lNode = xmssNode(skSeed, pkSeed, 2 * i, z - 1, adrs, params, hashSuite);
  const rNode = xmssNode(skSeed, pkSeed, 2 * i + 1, z - 1, adrs, params, hashSuite);

  adrs.setType(ADDR_TYPE.TREE);
  adrs.setTreeHeight(z);
  adrs.setTreeIndex(i);
  return hashSuite.H(pkSeed, adrs, lNode, rNode);
}

/**
 * xmssSign(M, skSeed, idx, pkSeed, adrs, params, hashSuite) - Algorithm 10
 * Generate an XMSS signature
 * Returns { sig, authPath } as concatenated bytes: sig || auth[0] || ... || auth[hp-1]
 */
function xmssSign(M, skSeed, idx, pkSeed, adrs, params, hashSuite) {
  const { hp, n } = params;

  // Generate WOTS+ signature
  adrs.setType(ADDR_TYPE.WOTS_HASH);
  adrs.setKeyPairAddress(idx);
  const sig = wotsSign(M, skSeed, pkSeed, adrs, params, hashSuite);

  // Build authentication path
  const auth = new Uint8Array(hp * n);
  for (let j = 0; j < hp; j++) {
    const k = (idx >>> j) ^ 1;  // sibling index at level j
    const node = xmssNode(skSeed, pkSeed, k, j, adrs, params, hashSuite);
    auth.set(node, j * n);
  }

  // Concatenate sig || auth
  const result = new Uint8Array(sig.length + auth.length);
  result.set(sig, 0);
  result.set(auth, sig.length);
  return result;
}

/**
 * xmssPkFromSig(idx, sigXmss, M, pkSeed, adrs, params, hashSuite) - Algorithm 11
 * Compute XMSS public key from signature
 */
function xmssPkFromSig(idx, sigXmss, M, pkSeed, adrs, params, hashSuite) {
  const { hp, n, len } = params;

  // Parse signature: first len*n bytes = WOTS sig, then hp*n bytes = auth path
  const wotsSig = sigXmss.slice(0, len * n);
  const auth = [];
  for (let j = 0; j < hp; j++) {
    auth.push(sigXmss.slice(len * n + j * n, len * n + (j + 1) * n));
  }

  // Compute WOTS+ public key from signature
  adrs.setType(ADDR_TYPE.WOTS_HASH);
  adrs.setKeyPairAddress(idx);
  let node0 = wotsPkFromSig(wotsSig, M, pkSeed, adrs, params, hashSuite);

  // Walk up the tree
  adrs.setType(ADDR_TYPE.TREE);
  adrs.setTreeIndex(idx);
  for (let j = 0; j < hp; j++) {
    adrs.setTreeHeight(j + 1);
    const treeIdx = (idx >>> j);
    if (treeIdx % 2 === 0) {
      adrs.setTreeIndex(treeIdx >>> 1);
      node0 = hashSuite.H(pkSeed, adrs, node0, auth[j]);
    } else {
      adrs.setTreeIndex((treeIdx - 1) >>> 1);
      node0 = hashSuite.H(pkSeed, adrs, auth[j], node0);
    }
  }
  return node0;
}

export { xmssNode, xmssSign, xmssPkFromSig };
